File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/widget/TimetableWidgetConfigureActivity.java
package de.tum.in.tumcampusapp.component.tumui.calendar.widget;

import android.appwidget.AppWidgetManager;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;
import android.widget.ListView;

import java.util.List;

import de.tum.in.tumcampusapp.R;
import de.tum.in.tumcampusapp.component.tumui.calendar.CalendarController;
import de.tum.in.tumcampusapp.component.tumui.calendar.model.CalendarItem;
import de.tum.in.tumcampusapp.component.tumui.lectures.adapter.LectureListSelectionAdapter;

public class TimetableWidgetConfigureActivity extends AppCompatActivity {

    private int appWidgetId;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_timetable_widget_configure);

        // Setup toolbar and save button
        setSupportActionBar(findViewById(R.id.main_toolbar));
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_check);

        // Get appWidgetId from intent
        Intent intent = getIntent();
        Bundle extras = intent.getExtras();
        if (extras != null) {
            appWidgetId = extras.getInt(
                    AppWidgetManager.EXTRA_APPWIDGET_ID,
                    AppWidgetManager.INVALID_APPWIDGET_ID);
        }

        ListView listViewLectures = findViewById(R.id.activity_timetable_lectures);

        // Initialize stations adapter
        CalendarController calendarController = new CalendarController(this);
        List<CalendarItem> lectures = calendarController.getLecturesForWidget(this.appWidgetId);
        listViewLectures.setAdapter(new LectureListSelectionAdapter(this, lectures, this.appWidgetId));
        listViewLectures.requestFocus();
    }

    /**
     * Setup cancel and back action
     *
     * @param item the menu item which has been pressed (or activated)
     */
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            saveAndReturn();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    /**
     * Saves the selection to the database, triggers a widget update and closes this activity
     */
    private void saveAndReturn() {
        // update widget
        Intent reloadIntent = new Intent(this, TimetableWidget.class);
        reloadIntent.setAction(TimetableWidget.BROADCAST_UPDATE_TIMETABLE_WIDGETS);
        reloadIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
        sendBroadcast(reloadIntent);

        // return to widget
        Intent resultValue = new Intent();
        resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
        setResult(RESULT_OK, resultValue);
        finish();
    }

}

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/widget/TimetableWidgetConfigureActivity.kt
package de.tum.`in`.tumcampusapp.component.tumui.calendar.widget

import android.app.Activity
import android.appwidget.AppWidgetManager
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.widget.ListView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import de.tum.`in`.tumcampusapp.R
import de.tum.`in`.tumcampusapp.component.tumui.calendar.CalendarController
import de.tum.`in`.tumcampusapp.component.tumui.lectures.adapter.LectureListSelectionAdapter

class TimetableWidgetConfigureActivity : AppCompatActivity() {

    private var appWidgetId: Int = 0

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_timetable_widget_configure)

        // Setup toolbar and save button
        setSupportActionBar(findViewById(R.id.toolbar))

        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        val closeIcon = ContextCompat.getDrawable(this, R.drawable.ic_check)
        if (closeIcon != null) {
            val color = ContextCompat.getColor(this, R.color.color_primary)
            closeIcon.setTint(color)
        }
        supportActionBar?.setHomeAsUpIndicator(closeIcon)

        // Get appWidgetId from intent
        appWidgetId = intent.extras?.getInt(
                AppWidgetManager.EXTRA_APPWIDGET_ID,
                AppWidgetManager.INVALID_APPWIDGET_ID) ?: 0

        val listViewLectures = findViewById<ListView>(R.id.activity_timetable_lectures)

        // Initialize stations adapter
        val calendarController = CalendarController(this)
        val lectures = calendarController.getLecturesForWidget(this.appWidgetId)
        listViewLectures.adapter = LectureListSelectionAdapter(this, lectures, this.appWidgetId)
        listViewLectures.requestFocus()
    }

    /**
     * Setup cancel and back action
     *
     * @param item the menu item which has been pressed (or activated)
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            saveAndReturn()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    /**
     * Saves the selection to the database, triggers a widget update and closes this activity
     */
    private fun saveAndReturn() {
        // update widget
        val reloadIntent = Intent(this, TimetableWidget::class.java).apply {
            action = TimetableWidget.BROADCAST_UPDATE_TIMETABLE_WIDGETS
            putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
        }
        sendBroadcast(reloadIntent)

        // return to widget
        val resultValue = Intent().apply {
            putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
        }
        setResult(Activity.RESULT_OK, resultValue)
        finish()
    }
}
